Starting point for sphinx-based user documentation

ejoerns opened the pull request:
This adds a basic skeleton for rauc-hawkbit-updater user documentation based on sphinx.

The tests are extended to test for doc (sphinx + doxygen) building, too.

Note that this also contains a fix for the current CI that fails to find changed debian repos.

The next step would then be to register a readthedocs subdomain for this, to allow browsing the documentation online.

Bastian-Krause commented on .github/workflows/tests.yml:
```diff
@@ -41,6 +42,25 @@ jobs:
+    - name: Install required packages
+      run: |
+        sudo apt-get update
+        sudo apt-get install libcurl4-openssl-dev libsystemd-dev libjson-glib-dev python-sphinx doxygen
```
`libsystemd-dev` shouldn't be needed here.

ejoerns replied:
Ack, removed

Bastian-Krause commented on CHANGES:
```diff
@@ -0,0 +1,4 @@
+Release 0.1 (in development)
+----------------------------
+
+This is the initial release of RAUC hakBit Updater.
```
```suggestion
This is the initial release of RAUC hawkBit Updater.
```

Bastian-Krause commented on docs/conf.py:
```diff
@@ -0,0 +1,274 @@
+# -*- coding: utf-8 -*-
+#
+# rauc-hawkbit-updater documentation build configuration file, created by
+# sphinx-quickstart on Fri Jan 22 16:00:15 2016.
```
Hm, better generate this with `sphinx-quickstart` instead? The generated version does not contain so much commented defaults and has better author etc. handling (less duplication).

ejoerns replied:
Yes, I re-created this now. I assumed the version I copied from rauc was also the result form a `sphinx-quickstart` run, but most of the things in there we do not really need. The new version looks much shorter.

Bastian-Krause commented on docs/contributing.rst:
```diff
@@ -0,0 +1,76 @@
+============
+
+Thank you for thinking about contributing to RAUC hawkBit Updater!
+Some different backgrounds and use-cases are essential for making RAUC hawkBit
```
```suggestion
Various backgrounds and use-cases are essential for making RAUC hawkBit
```

Bastian-Krause commented on docs/contributing.rst:
```diff
@@ -0,0 +1,76 @@
+- Basically follow the Linux kernel coding style
+
+Documentation
+-------------
+- Use `semantic linefeeds
```
```suggestion
-------------

- Use `semantic linefeeds
```

Bastian-Krause commented on docs/reference.rst:
```diff
@@ -0,0 +1,113 @@
+Configuration File
+------------------
+
+Example configuration:
+
+.. code-block:: cfg
+
+  [client]
+  hawkbit_server            = 127.0.0.1:8080
+  ssl                       = false
+  ssl_verify                = false
+  tenant_id                 = DEFAULT
+  target_name               = test-target
+  auth_token                = bhVahL1Il1shie2aj2poojeChee6ahShu
+  #gateway_token            = bhVahL1Il1shie2aj2poojeChee6ahShu
+  bundle_download_location  = /tmp/bundle.raucb
+  retry_wait                = 60
+  connect_timeout           = 20
+  timeout                   = 60
+  log_level                 = debug
+
+  [device]
+  product                   = Terminator
+  model                     = T-1000
+  serialnumber              = 8922673153
+  hw_revision               = 2
+  key1                      = value
+  key2                      = value
```
Do we need this here? This is the same as config.conf.example in the project's root.

ejoerns replied:
I'd tend to keep it as people reading the documentation online might not have the source present. I'll shorten this a little

Bastian-Krause commented on docs/reference.rst:
```diff
@@ -0,0 +1,113 @@
+
+**[client] section**
+
+Configures how to connect with the hawkBit server, etc.
```
```suggestion
Configures how to connect to a hawkBit server, etc.
```

Bastian-Krause commented on docs/reference.rst:
```diff
@@ -0,0 +1,113 @@
+Mandatory options:
+
+``hawkbit_server=<host>[:<port>]``
+  The ip or hostname of the hawkbitServer to connect to.
```
```suggestion
  The IP or hostname of the hawkBit server to connect to.
```

Bastian-Krause commented on docs/reference.rst:
```diff
@@ -0,0 +1,113 @@
+
+``hawkbit_server=<host>[:<port>]``
+  The ip or hostname of the hawkbitServer to connect to.
+  The ``port`` can be provided optionally.
+  It needs to be separated with colons.
```
```suggestion
  The ``port`` can be provided optionally, separated by a colon.
```

Bastian-Krause commented on docs/reference.rst:
```diff
@@ -0,0 +1,113 @@
+
+``gateway_token``
+  Gateway authentication token.
+  This is set by hawkBit.
```
What does this mean? This can be set just like the authentication token on hawkBit side (via the [managament API](https://www.eclipse.org/hawkbit/apis/management_api/)) or generated randomly by hawkBit. Maybe better mention that this is a tenant-wide token and it must be enabled first?

ejoerns replied:
You're right. I've tried to be more specific here and also linked the hawkbit doc for authentication.

Bastian-Krause commented on docs/reference.rst:
```diff
@@ -0,0 +1,113 @@
+``bundle_download_location=<path>``
+  Full path to where the bundle should be downloaded to.
+  E.g. set to ``/tmp/_bundle.raucb`` to let rauc-hawkbit-updater use this
+  location withing ``/tmp``.
```
```suggestion
  location within ``/tmp``.
```

Bastian-Krause commented on docs/reference.rst:
```diff
@@ -0,0 +1,113 @@
+Optional options:
+  
+``tenant_id=<ID>``
+  ID of the tenant to connect with. Defaults to ``DEFAULT``.
```
```suggestion
  ID of the tenant to connect to. Defaults to ``DEFAULT``.
```

Bastian-Krause commented on docs/reference.rst:
```diff
@@ -0,0 +1,113 @@
+  Defaults to 20 seconds.
+
+``timeout=<seconds>``
+  HTTP request Timeout [seconds]
```
```suggestion
  HTTP request timeout [seconds].
```

Bastian-Krause commented on docs/reference.rst:
```diff
@@ -0,0 +1,113 @@
+  Defaults to 60 seconds.
+
+``retry_wait=<seconds>``
+  Time in seconds to wait before retrying [seconds]
```
```suggestion
  Time to wait before retrying in case an error occurred [seconds].
```

Bastian-Krause reviewed (changes requested):
Thanks!

`docs/extensions/lexer/DtsLexer.pyc` shouldn't be under version control. Maybe add this to .gitignore?

ejoerns commented:
I've rebased this on master so the CI fix is now not part of this anymore.

Additionally, I structured it in a way that it starts from a vanilla sphinx-quickstart-generated configuration.

The theme is now changed to use readthedocs which requires an additional package but will have the same appearance in local builds as it would have for readthedocs builds.

Bastian-Krause commented on docs/using.rst:
```diff
@@ -0,0 +1,23 @@
+Using The RAUC hawkbit Updater
```
```suggestion
Using the RAUC hawkbit Updater
```

Bastian-Krause commented on docs/using.rst:
```diff
@@ -0,0 +1,23 @@
+Authentication
+--------------
+
+As described on the [hawkBit Authentication page](https://www.eclipse.org/hawkbit/concepts/authentication/)
```
```suggestion
As described on the `hawkBit Authentication page <https://www.eclipse.org/hawkbit/concepts/authentication/>`_
```

Bastian-Krause commented on docs/using.rst:
```diff
@@ -0,0 +1,23 @@
+with a security token. A security token can be either a "Target" token or a
+"Gateway" token. The "Target" security token is specific to a single target
+defined in hawkBit. In the RAUC hawkBit updater's configuration file it's
+referred to as "auth_token".
```
```suggestion
referred to as ``auth_token``.
```

Bastian-Krause commented on docs/using.rst:
```diff
@@ -0,0 +1,23 @@
+allow it to manage all the targets under its tenant. With RAUC hawkBit updater
+such token can be used to authenticate all targets on the server. I.e. same
+gateway token can be used in a configuration file replicated on many targets.
+In the RAUC hawkBit updater's configuration file it's called "gateway_token".
```
```suggestion
In the RAUC hawkBit updater's configuration file it's called ``gateway_token``.
```

ejoerns commented:
@Bastian-Krause I should have finally addressed your review comments now.

ejoerns commented:
Rebased to benefit from testing fix #76.

Bastian-Krause reviewed (approved):
Nice!

Bastian-Krause commented:
@ejoerns Pushed some small fixups, could you have a look and squash them if they look good?

ejoerns commented:
@Bastian-Krause Thanks, they look quite reasonable! I've fixed up all but the one that adds new text and already got a SoB anyway. I left this as a stand-alone commit.
